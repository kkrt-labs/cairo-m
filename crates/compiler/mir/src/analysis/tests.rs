//! # Tests for Analysis Module
//!
//! This module contains tests for dominance analysis and other SSA-related analyses.

use super::dominance::{compute_dominance_frontiers, compute_dominator_tree};
use crate::{BasicBlockId, MirFunction, Terminator, Value};
use rustc_hash::FxHashSet;

/// Helper to create a simple linear CFG: Entry -> B1 -> B2 -> Exit
fn create_linear_cfg() -> MirFunction {
    let mut function = MirFunction::new("test_linear".to_string());

    // Create 4 blocks
    for _ in 0..4 {
        function.basic_blocks.push(crate::BasicBlock::new());
    }

    // Entry (0) -> B1 (1)
    function.basic_blocks[BasicBlockId::from_raw(0)].terminator = Terminator::Jump {
        target: BasicBlockId::from_raw(1),
    };

    // B1 (1) -> B2 (2)
    function.basic_blocks[BasicBlockId::from_raw(1)].terminator = Terminator::Jump {
        target: BasicBlockId::from_raw(2),
    };

    // B2 (2) -> Exit (3)
    function.basic_blocks[BasicBlockId::from_raw(2)].terminator = Terminator::Jump {
        target: BasicBlockId::from_raw(3),
    };

    // Exit (3) returns
    function.basic_blocks[BasicBlockId::from_raw(3)].terminator =
        Terminator::Return { values: vec![] };

    function
}

/// Helper to create an if-else diamond CFG:
///     Entry
///     /  \
///   Then  Else
///     \  /
///     Merge
fn create_if_else_cfg() -> MirFunction {
    let mut function = MirFunction::new("test_if_else".to_string());

    // Create 4 blocks: Entry, Then, Else, Merge
    for _ in 0..4 {
        function.basic_blocks.push(crate::BasicBlock::new());
    }

    let entry = BasicBlockId::from_raw(0);
    let then_block = BasicBlockId::from_raw(1);
    let else_block = BasicBlockId::from_raw(2);
    let merge = BasicBlockId::from_raw(3);

    // Entry branches to Then or Else
    let cond = function.new_value_id();
    function.basic_blocks[entry].terminator = Terminator::If {
        condition: Value::operand(cond),
        then_target: then_block,
        else_target: else_block,
    };

    // Then -> Merge
    function.basic_blocks[then_block].terminator = Terminator::Jump { target: merge };

    // Else -> Merge
    function.basic_blocks[else_block].terminator = Terminator::Jump { target: merge };

    // Merge returns
    function.basic_blocks[merge].terminator = Terminator::Return { values: vec![] };

    function
}

/// Helper to create a simple loop CFG:
///     Entry
///       |
///     Header <--
///       |      |
///     Body ----
///       |
///     Exit
fn create_loop_cfg() -> MirFunction {
    let mut function = MirFunction::new("test_loop".to_string());

    // Create 4 blocks: Entry, Header, Body, Exit
    for _ in 0..4 {
        function.basic_blocks.push(crate::BasicBlock::new());
    }

    let entry = BasicBlockId::from_raw(0);
    let header = BasicBlockId::from_raw(1);
    let body = BasicBlockId::from_raw(2);
    let exit = BasicBlockId::from_raw(3);

    // Entry -> Header
    function.basic_blocks[entry].terminator = Terminator::Jump { target: header };

    // Header branches to Body or Exit
    let cond = function.new_value_id();
    function.basic_blocks[header].terminator = Terminator::If {
        condition: Value::operand(cond),
        then_target: body,
        else_target: exit,
    };

    // Body -> Header (loop back)
    function.basic_blocks[body].terminator = Terminator::Jump { target: header };

    // Exit returns
    function.basic_blocks[exit].terminator = Terminator::Return { values: vec![] };

    function
}

/// Helper to create a nested if CFG:
///        Entry
///        /  \
///      B1    B2
///     / \     |
///   B3  B4   /
///     \ |  /
///      Merge
fn create_nested_if_cfg() -> MirFunction {
    let mut function = MirFunction::new("test_nested_if".to_string());

    // Create 6 blocks
    for _ in 0..6 {
        function.basic_blocks.push(crate::BasicBlock::new());
    }

    let entry = BasicBlockId::from_raw(0);
    let b1 = BasicBlockId::from_raw(1);
    let b2 = BasicBlockId::from_raw(2);
    let b3 = BasicBlockId::from_raw(3);
    let b4 = BasicBlockId::from_raw(4);
    let merge = BasicBlockId::from_raw(5);

    // Entry branches to B1 or B2
    let cond1 = function.new_value_id();
    function.basic_blocks[entry].terminator = Terminator::If {
        condition: Value::operand(cond1),
        then_target: b1,
        else_target: b2,
    };

    // B1 branches to B3 or B4
    let cond2 = function.new_value_id();
    function.basic_blocks[b1].terminator = Terminator::If {
        condition: Value::operand(cond2),
        then_target: b3,
        else_target: b4,
    };

    // B2 -> Merge
    function.basic_blocks[b2].terminator = Terminator::Jump { target: merge };

    // B3 -> Merge
    function.basic_blocks[b3].terminator = Terminator::Jump { target: merge };

    // B4 -> Merge
    function.basic_blocks[b4].terminator = Terminator::Jump { target: merge };

    // Merge returns
    function.basic_blocks[merge].terminator = Terminator::Return { values: vec![] };

    function
}

#[test]
fn test_dominator_tree_linear() {
    let function = create_linear_cfg();
    let dom_tree = compute_dominator_tree(&function);

    // In a linear CFG, each block is dominated by its predecessor
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(1)],
        BasicBlockId::from_raw(0)
    );
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(2)],
        BasicBlockId::from_raw(1)
    );
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(3)],
        BasicBlockId::from_raw(2)
    );
}

#[test]
fn test_dominator_tree_if_else() {
    let function = create_if_else_cfg();
    let dom_tree = compute_dominator_tree(&function);

    // Entry dominates all blocks
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(1)],
        BasicBlockId::from_raw(0)
    ); // Then
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(2)],
        BasicBlockId::from_raw(0)
    ); // Else
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(3)],
        BasicBlockId::from_raw(0)
    ); // Merge
}

#[test]
fn test_dominator_tree_loop() {
    let function = create_loop_cfg();
    let dom_tree = compute_dominator_tree(&function);

    // Entry dominates Header
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(1)],
        BasicBlockId::from_raw(0)
    );

    // Header dominates Body and Exit
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(2)],
        BasicBlockId::from_raw(1)
    );
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(3)],
        BasicBlockId::from_raw(1)
    );
}

#[test]
fn test_dominator_tree_nested_if() {
    let function = create_nested_if_cfg();
    let dom_tree = compute_dominator_tree(&function);

    // Entry dominates B1 and B2
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(1)],
        BasicBlockId::from_raw(0)
    );
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(2)],
        BasicBlockId::from_raw(0)
    );

    // B1 dominates B3 and B4
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(3)],
        BasicBlockId::from_raw(1)
    );
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(4)],
        BasicBlockId::from_raw(1)
    );

    // Entry dominates Merge
    assert_eq!(
        dom_tree[&BasicBlockId::from_raw(5)],
        BasicBlockId::from_raw(0)
    );
}

#[test]
fn test_dominance_frontiers_linear() {
    let function = create_linear_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    // In a linear CFG, no block has a dominance frontier
    for i in 0..4 {
        assert!(frontiers[&BasicBlockId::from_raw(i)].is_empty());
    }
}

#[test]
fn test_dominance_frontiers_if_else() {
    let function = create_if_else_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    // Entry has no frontier
    assert!(frontiers[&BasicBlockId::from_raw(0)].is_empty());

    // Then and Else blocks have Merge in their frontier
    // (They dominate predecessors of Merge but not Merge itself)
    assert_eq!(
        frontiers[&BasicBlockId::from_raw(1)],
        FxHashSet::from_iter([BasicBlockId::from_raw(3)])
    );
    assert_eq!(
        frontiers[&BasicBlockId::from_raw(2)],
        FxHashSet::from_iter([BasicBlockId::from_raw(3)])
    );

    // Merge has no frontier
    assert!(frontiers[&BasicBlockId::from_raw(3)].is_empty());
}

#[test]
fn test_dominance_frontiers_loop() {
    let function = create_loop_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    // Entry has no frontier
    assert!(frontiers[&BasicBlockId::from_raw(0)].is_empty());

    // Header has itself in its frontier (because Body dominates a predecessor of Header)
    assert_eq!(
        frontiers[&BasicBlockId::from_raw(1)],
        FxHashSet::from_iter([BasicBlockId::from_raw(1)])
    );

    // Body has Header in its frontier (loop back edge)
    assert_eq!(
        frontiers[&BasicBlockId::from_raw(2)],
        FxHashSet::from_iter([BasicBlockId::from_raw(1)])
    );

    // Exit has no frontier
    assert!(frontiers[&BasicBlockId::from_raw(3)].is_empty());
}

#[test]
fn test_dominance_frontiers_nested_if() {
    let function = create_nested_if_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    // Entry has no frontier
    assert!(frontiers[&BasicBlockId::from_raw(0)].is_empty());

    // B1 has Merge in its frontier (because its children B3/B4 dominate predecessors of Merge)
    assert_eq!(
        frontiers[&BasicBlockId::from_raw(1)],
        FxHashSet::from_iter([BasicBlockId::from_raw(5)])
    );

    // B2, B3, and B4 all have Merge in their frontier
    assert_eq!(
        frontiers[&BasicBlockId::from_raw(2)],
        FxHashSet::from_iter([BasicBlockId::from_raw(5)])
    );
    assert_eq!(
        frontiers[&BasicBlockId::from_raw(3)],
        FxHashSet::from_iter([BasicBlockId::from_raw(5)])
    );
    assert_eq!(
        frontiers[&BasicBlockId::from_raw(4)],
        FxHashSet::from_iter([BasicBlockId::from_raw(5)])
    );

    // Merge has no frontier
    assert!(frontiers[&BasicBlockId::from_raw(5)].is_empty());
}

#[test]
fn test_multiple_returns() {
    let mut function = MirFunction::new("test_multiple_returns".to_string());

    // Create 3 blocks: Entry, Return1, Return2
    for _ in 0..3 {
        function.basic_blocks.push(crate::BasicBlock::new());
    }

    let entry = BasicBlockId::from_raw(0);
    let ret1 = BasicBlockId::from_raw(1);
    let ret2 = BasicBlockId::from_raw(2);

    // Entry branches to Return1 or Return2
    let cond = function.new_value_id();
    function.basic_blocks[entry].terminator = Terminator::If {
        condition: Value::operand(cond),
        then_target: ret1,
        else_target: ret2,
    };

    // Both blocks return
    function.basic_blocks[ret1].terminator = Terminator::Return { values: vec![] };
    function.basic_blocks[ret2].terminator = Terminator::Return { values: vec![] };

    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    // Entry dominates both returns
    assert_eq!(dom_tree[&ret1], entry);
    assert_eq!(dom_tree[&ret2], entry);

    // No block has a dominance frontier (no merge points)
    assert!(frontiers[&entry].is_empty());
    assert!(frontiers[&ret1].is_empty());
    assert!(frontiers[&ret2].is_empty());
}

/// Helper to create an irreducible CFG (multiple entry points to a loop)
///     Entry
///     /  \
///   B1    B2
///     \  / \
///      B3--B4
fn create_irreducible_cfg() -> MirFunction {
    let mut function = MirFunction::new("test_irreducible".to_string());

    // Create 5 blocks
    for _ in 0..5 {
        function.basic_blocks.push(crate::BasicBlock::new());
    }

    let entry = BasicBlockId::from_raw(0);
    let b1 = BasicBlockId::from_raw(1);
    let b2 = BasicBlockId::from_raw(2);
    let b3 = BasicBlockId::from_raw(3);
    let b4 = BasicBlockId::from_raw(4);

    // Entry branches to B1 or B2
    let cond1 = function.new_value_id();
    function.basic_blocks[entry].terminator = Terminator::If {
        condition: Value::operand(cond1),
        then_target: b1,
        else_target: b2,
    };

    // B1 -> B3
    function.basic_blocks[b1].terminator = Terminator::Jump { target: b3 };

    // B2 branches to B3 or B4
    let cond2 = function.new_value_id();
    function.basic_blocks[b2].terminator = Terminator::If {
        condition: Value::operand(cond2),
        then_target: b3,
        else_target: b4,
    };

    // B3 -> B4 (creates a cycle with B4)
    function.basic_blocks[b3].terminator = Terminator::Jump { target: b4 };

    // B4 -> B3 (back edge creating irreducible loop)
    function.basic_blocks[b4].terminator = Terminator::Jump { target: b3 };

    function
}

#[test]
fn test_irreducible_cfg() {
    let function = create_irreducible_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    let entry = BasicBlockId::from_raw(0);
    let b1 = BasicBlockId::from_raw(1);
    let b2 = BasicBlockId::from_raw(2);
    let b3 = BasicBlockId::from_raw(3);
    let b4 = BasicBlockId::from_raw(4);

    // Entry dominates B1 and B2
    assert_eq!(dom_tree[&b1], entry);
    assert_eq!(dom_tree[&b2], entry);

    // Entry dominates B3 and B4 (not B1 or B2 since there are multiple paths)
    assert_eq!(dom_tree[&b3], entry);
    assert_eq!(dom_tree[&b4], entry);

    // B3 and B4 form an irreducible loop - they are in each other's DF
    assert!(frontiers[&b3].contains(&b4));
    assert!(frontiers[&b4].contains(&b3));

    // B3 is a join point (from B1 and B2)
    assert!(frontiers[&b1].contains(&b3) || frontiers[&b2].contains(&b3));
}

/// Helper to create a complex nested loop CFG
///     Entry
///       |
///     Outer <----
///     /  \      |
///   Inner Exit  |
///     |\ /      |
///     | X       |
///     |/ \      |
///   Back--+-----
fn create_nested_loops_cfg() -> MirFunction {
    let mut function = MirFunction::new("test_nested_loops".to_string());

    // Create 5 blocks: Entry, Outer, Inner, Back, Exit
    for _ in 0..5 {
        function.basic_blocks.push(crate::BasicBlock::new());
    }

    let entry = BasicBlockId::from_raw(0);
    let outer = BasicBlockId::from_raw(1);
    let inner = BasicBlockId::from_raw(2);
    let back = BasicBlockId::from_raw(3);
    let exit = BasicBlockId::from_raw(4);

    // Entry -> Outer
    function.basic_blocks[entry].terminator = Terminator::Jump { target: outer };

    // Outer branches to Inner or Exit
    let cond1 = function.new_value_id();
    function.basic_blocks[outer].terminator = Terminator::If {
        condition: Value::operand(cond1),
        then_target: inner,
        else_target: exit,
    };

    // Inner branches to Back or Exit
    let cond2 = function.new_value_id();
    function.basic_blocks[inner].terminator = Terminator::If {
        condition: Value::operand(cond2),
        then_target: back,
        else_target: exit,
    };

    // Back branches to Inner (inner loop) or Outer (outer loop)
    let cond3 = function.new_value_id();
    function.basic_blocks[back].terminator = Terminator::If {
        condition: Value::operand(cond3),
        then_target: inner,
        else_target: outer,
    };

    // Exit returns
    function.basic_blocks[exit].terminator = Terminator::Return { values: vec![] };

    function
}

#[test]
fn test_nested_loops() {
    let function = create_nested_loops_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    let entry = BasicBlockId::from_raw(0);
    let outer = BasicBlockId::from_raw(1);
    let inner = BasicBlockId::from_raw(2);
    let back = BasicBlockId::from_raw(3);
    let exit = BasicBlockId::from_raw(4);

    // Verify dominator tree
    assert_eq!(dom_tree[&outer], entry);
    assert_eq!(dom_tree[&inner], outer);
    assert_eq!(dom_tree[&back], inner);
    assert_eq!(dom_tree[&exit], outer);

    // Outer is in its own DF (loop header for outer loop)
    assert!(frontiers[&outer].contains(&outer));

    // Inner is in its own DF (loop header for inner loop)
    assert!(frontiers[&inner].contains(&inner));

    // Back node should have both loop headers in its DF
    assert!(frontiers[&back].contains(&inner));
    assert!(frontiers[&back].contains(&outer));

    // Exit is a join point from Outer and Inner
    assert!(frontiers[&inner].contains(&exit));
}

/// Helper to create a diamond with critical edge
///     Entry
///       |
///      B1
///     /  \
///   B2    B3
///   |  X  |
///   | / \ |
///   B4   B5
///    \ /
///    Exit
fn create_critical_edge_cfg() -> MirFunction {
    let mut function = MirFunction::new("test_critical_edge".to_string());

    // Create 7 blocks
    for _ in 0..7 {
        function.basic_blocks.push(crate::BasicBlock::new());
    }

    let entry = BasicBlockId::from_raw(0);
    let b1 = BasicBlockId::from_raw(1);
    let b2 = BasicBlockId::from_raw(2);
    let b3 = BasicBlockId::from_raw(3);
    let b4 = BasicBlockId::from_raw(4);
    let b5 = BasicBlockId::from_raw(5);
    let exit = BasicBlockId::from_raw(6);

    // Entry -> B1
    function.basic_blocks[entry].terminator = Terminator::Jump { target: b1 };

    // B1 branches to B2 or B3
    let cond1 = function.new_value_id();
    function.basic_blocks[b1].terminator = Terminator::If {
        condition: Value::operand(cond1),
        then_target: b2,
        else_target: b3,
    };

    // B2 branches to B4 or B5 (critical edges to join points)
    let cond2 = function.new_value_id();
    function.basic_blocks[b2].terminator = Terminator::If {
        condition: Value::operand(cond2),
        then_target: b4,
        else_target: b5,
    };

    // B3 branches to B4 or B5 (critical edges to join points)
    let cond3 = function.new_value_id();
    function.basic_blocks[b3].terminator = Terminator::If {
        condition: Value::operand(cond3),
        then_target: b4,
        else_target: b5,
    };

    // B4 -> Exit
    function.basic_blocks[b4].terminator = Terminator::Jump { target: exit };

    // B5 -> Exit
    function.basic_blocks[b5].terminator = Terminator::Jump { target: exit };

    // Exit returns
    function.basic_blocks[exit].terminator = Terminator::Return { values: vec![] };

    function
}

#[test]
fn test_critical_edges() {
    let function = create_critical_edge_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    let b1 = BasicBlockId::from_raw(1);
    let b2 = BasicBlockId::from_raw(2);
    let b3 = BasicBlockId::from_raw(3);
    let b4 = BasicBlockId::from_raw(4);
    let b5 = BasicBlockId::from_raw(5);
    let exit = BasicBlockId::from_raw(6);

    // B4 and B5 are join points with critical edges
    // They should be in the DF of B2 and B3
    assert!(frontiers[&b2].contains(&b4));
    assert!(frontiers[&b2].contains(&b5));
    assert!(frontiers[&b3].contains(&b4));
    assert!(frontiers[&b3].contains(&b5));

    // Exit is a join point from B4 and B5
    assert!(frontiers[&b4].contains(&exit));
    assert!(frontiers[&b5].contains(&exit));

    // Verify dominator relationships
    assert_eq!(dom_tree[&b4], b1);
    assert_eq!(dom_tree[&b5], b1);
    assert_eq!(dom_tree[&exit], b1);
}

#[test]
fn test_phi_placement_correctness() {
    // Test that phi nodes would be placed correctly for a variable defined in multiple branches
    // This mimics what mem2reg does with the DF computation
    let function = create_if_else_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    // Simulate placing phi nodes for a variable defined in blocks 1 and 2
    let mut phi_blocks = FxHashSet::default();
    let def_blocks = vec![BasicBlockId::from_raw(1), BasicBlockId::from_raw(2)];

    // Standard algorithm: for each def block, add its DF to phi blocks
    for &def_block in &def_blocks {
        for &df_block in &frontiers[&def_block] {
            phi_blocks.insert(df_block);
        }
    }

    // Phi should be placed at the merge block (block 3)
    assert_eq!(phi_blocks.len(), 1);
    assert!(phi_blocks.contains(&BasicBlockId::from_raw(3)));
}

#[test]
fn test_loop_phi_placement() {
    // Test phi placement for loop variables
    let function = create_loop_cfg();
    let dom_tree = compute_dominator_tree(&function);
    let frontiers = compute_dominance_frontiers(&function, &dom_tree);

    // Simulate placing phi nodes for a variable modified in the loop body
    let mut phi_blocks = FxHashSet::default();
    let def_blocks = vec![BasicBlockId::from_raw(2)]; // Body modifies variable

    for &def_block in &def_blocks {
        for &df_block in &frontiers[&def_block] {
            phi_blocks.insert(df_block);
        }
    }

    // Phi should be placed at the loop header (block 1)
    assert!(phi_blocks.contains(&BasicBlockId::from_raw(1)));
}
