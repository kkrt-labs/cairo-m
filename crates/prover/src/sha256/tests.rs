use num_traits::Zero;
use stwo_prover::core::fields::m31::M31;
use stwo_prover::core::vcs::blake2_merkle::Blake2sMerkleChannel;

use crate::adapter::SHA256HashInput;
use crate::sha256::debug_tools::assert_constraints::assert_constraints;
use crate::sha256::prover_sha256::prove_sha256;

#[test]
fn test_sha256_constraints_empty_input() {
    let mut inputs: Vec<SHA256HashInput> = vec![];
    assert_constraints(&mut inputs);
}

#[test]
fn test_sha256_constraints_single_block() {
    // Create a single SHA256 block input (32 M31 elements = 512 bits)
    let mut inputs: Vec<SHA256HashInput> = vec![[M31::from(42); 32]];
    assert_constraints(&mut inputs);
}

#[test]
fn test_sha256_prove_empty_input() {
    let inputs: Vec<SHA256HashInput> = vec![];
    prove_sha256::<Blake2sMerkleChannel>(&inputs, None).unwrap();
}

#[test]
fn test_sha256_prove_single_block() {
    let inputs: Vec<SHA256HashInput> = vec![[M31::from(42); 32]];
    prove_sha256::<Blake2sMerkleChannel>(&inputs, None).unwrap();
}

#[test]
fn test_sha256_prove_2_pow_x_input() {
    let inputs: Vec<SHA256HashInput> = (0..1 << 16)
        .map(|i| {
            let mut message = [M31::zero(); 32];
            for (j, element) in message.iter_mut().enumerate() {
                *element = M31::from((i + j) as u32 & 0xFFFF);
            }
            message
        })
        .collect();
    prove_sha256::<Blake2sMerkleChannel>(&inputs, None).unwrap();
}
