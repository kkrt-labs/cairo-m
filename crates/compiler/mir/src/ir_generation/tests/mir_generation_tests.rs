//! Main test runner for MIR generation.

use insta::assert_snapshot;

use crate::ir_generation::tests::test_harness::{MirTest, check_mir};

/// A macro to define a MIR test case. It loads a source file,
/// runs MIR generation, checks embedded assertions, and snapshots the output.
macro_rules! mir_test {
    ($test_name:ident, $subdir:expr) => {
        #[test]
        fn $test_name() {
            // Construct the path to the test source file.
            let path = concat!("test_cases/", $subdir, "/", stringify!($test_name), ".cm");

            // Load the test case, which also parses assertions.
            let test = MirTest::load(path);

            // Generate MIR from the source code.
            let mir_output = check_mir(&test.source);


            // Use insta to snapshot the entire MIR test output.
            // The snapshot name is derived from the subdirectory and test name for clarity.
            let snapshot_content = format!(
                "---\nsource: {}\nexpression: mir_output\n---\nFixture: {}.cm\n============================================================\nSource code:\n{}\n============================================================\nGenerated MIR:\n{}",
                file!(),
                stringify!($test_name),
                test.source,
                mir_output.mir_string
            );
            assert_snapshot!(concat!($subdir, "_", stringify!($test_name)), snapshot_content);

            // Validate assertions after snapshotting (to get visual output).
            test.check_assertions(&mir_output);
        }
    };
}

// ====== Test Groups ======

// --- Simple Functions ---
mir_test!(function_simple, "simple");
mir_test!(function_with_let, "simple");
mir_test!(function_with_params, "simple");

// --- Expressions ---
mir_test!(unary_ops, "expressions");
mir_test!(binary_ops, "expressions");
mir_test!(comparison_ops, "expressions");
mir_test!(compound_expr, "expressions");
mir_test!(operator_precedence, "expressions");

// --- Control Flow ---
mir_test!(if_else, "control_flow");
mir_test!(if_partial_return, "control_flow");
mir_test!(if_both_return, "control_flow");
mir_test!(if_no_else, "control_flow");
mir_test!(nested_if, "control_flow");
mir_test!(unreachable_after_return, "control_flow");
mir_test!(simple_while, "control_flow");
mir_test!(infinite_loop, "control_flow");
mir_test!(nested_loops, "control_flow");
mir_test!(loop_with_breaks, "control_flow");

// --- Functions ---
mir_test!(call_as_statement, "functions");
mir_test!(call_with_return, "functions");
mir_test!(multiple_functions, "functions");
mir_test!(fib, "functions");
mir_test!(fib_loop, "functions");
mir_test!(return_values, "functions");

// --- Variables ---
mir_test!(assignment, "variables");
mir_test!(reassignment_from_var, "variables");

// --- Optimizations ---
mir_test!(unused_variable_elimination, "optimizations");

// --- Aggregates (Structs/Tuples) ---
mir_test!(struct_literal, "aggregates");
mir_test!(struct_access_mut, "aggregates");
mir_test!(struct_access_read, "aggregates");
mir_test!(tuple_literal_and_access, "aggregates");
// TODO this theoritically works but we dont have proper memory allocation mechanisms.
mir_test!(array_access, "aggregates");

// --- Tuple Destructuring ---
mir_test!(tuple_destructuring, "expressions");
