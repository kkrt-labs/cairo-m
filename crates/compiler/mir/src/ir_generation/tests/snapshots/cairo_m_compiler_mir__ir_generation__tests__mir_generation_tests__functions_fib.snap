---
source: crates/compiler/mir/src/ir_generation/tests/mir_generation_tests.rs
expression: snapshot_content
---
---
source: crates/compiler/mir/src/ir_generation/tests/mir_generation_tests.rs
expression: mir_output
---
Fixture: fib.cm
============================================================
Source code:
fn main() -> felt {
    let n = 10;
    let result = fib(n);
    return result;
}

fn fib(n: felt) -> felt {
    if (n == 0) {
        return 0;
    }
    if (n == 1) {
        return 1;
    }
    let res1 = fib(n - 1);
    let res2 = fib(n - 2);
    return res1 + res2;
}

============================================================
Generated MIR:
module {
  // Function 0
  fn main {
    entry: 0

    0:
      %0 = stackalloc 1
      store %0, 10
      %1 = call 1(%0)
      %2 = stackalloc 1
      store %2, %1
      return %2

  }

  // Function 1
  fn fib {
    parameters: [0]
    entry: 0

    0:
      if %0 then jump 2 else jump 1

    1:
      return 0

    2:
      if %0 Eq 1 then jump 3 else jump 4

    3:
      return 1

    4:
      %5 = %0 Sub 1
      %6 = call 1(%5)
      %7 = stackalloc 1
      store %7, %6
      %8 = %0 Sub 2
      %9 = call 1(%8)
      %10 = stackalloc 1
      store %10, %9
      %11 = %7 Add %10
      return %11

  }

}
